[e-m-c][Android] Create `ActivityResultsRegistry`

bbarthec opened the pull request:
# Why

Extracted part of #16251
Depends on #17571 

Android is proposing a new way for registering for activity results - using [`ComponentActivity.registerForActivityResult`]( https://developer.android.com/training/basics/intents/result).

The old way of using `startActivityForResult` and listening for results `onActivityResult` is deprecated.

Even though https://github.com/expo/expo/pull/17571 would enforce the underlying Activity to support this new mechanism we cannot use it directly on Activity. 
`react-native`'s application lifecycle is not mirroring the Android lifecycle (and thus Activity's lifecycle) completely. The moment the `ComponentActivity` expects the registration to happen is not available for us neither in `expo-modules-core` neither in `react-native`'s application lifecycle. What is more these two lifecycle lives independently to each other.

In order to be able to use modern Android libraries (e.g. https://github.com/CanHub/Android-Image-Cropper) we need to support this new mechanism somehow. 

I've copied and adjusted the original `AndroidX` files in order to couple the lifecycle of registered contracts not with `Activity`, but with `AppContext` (and thus `bridge`) lifecycle. 

# How

I've copied `ActivityResultsManager`, `ActivityResultRegistry` and `ActivityResultCallback` from `androidx` libraries and adjusted them to match `AppContext` lifecycle:
- as `AppContext` is alive as long as the application is alive, we don't have to save data in `onSaveInstanceState` and retrieve it from `onCreate` (`rn-screens` and `react-navigation` are even strictly forcing us to mark the saved state as `null` when `onCreate` is called in order to work properly)
- `registerForActivityResults` mechanism is based on two different callbacks:
    - `main` callback that is passed when `launch` is called on `ActivityResultLauncher` - this one could be called stateful callback, because it has access to ReactNative objects, like `Promise`, etc
    - `fallback` callback that is registered at the very beginning of the module's lifecycle. It handles all the data that is returned to the application, but only when the `Activity` has been killed and recreated by the system. When this happens `react-native` recreates the whole application from scratch, so each module has to store possible orphaned data and provide a way to retrieve them.
      - right now we have: https://docs.expo.dev/versions/latest/sdk/imagepicker/#imagepickergetpendingresultasync
      - Flutter has similar mechanism: https://pub.dev/documentation/image_picker/latest/#handling-mainactivity-destruction-on-android

Additionally I had to modify a bit how ReactNative propagates `onActivityResult` - see https://github.com/expo/expo/pull/17572/commits/150c5e7ba6277c52eac14de75e86b4afe4bef50f. This change has to be discussed and properly applied to all projects using `expo-modules-core`. I'm not sure how `ReactNativeDelegateWrapper` is applied to every project, so I'm asking for a bit of help with this aspect 🙏 👀 

# Test Plan

To be testable with #17671 
- [x] I've checked flows for both `Images` and `Videos`
- [x] I've also checked the main bullet point of this mechanism - how it behaves when `Activity` is being recreated by Android - there's a flag in `Developer Options` available on the device that simulates such scenario (`Don't keep Activities`)

bbarthec committed:
[emc][Android] Add `ActivityResultsRegistry`

`ReactNative` apps that uses `emc` cannot hook propoerly into `Activity` lifecycle and thus we cannot cannot directly use all fancy stuff from `AndroidX`.
I'm proposing similar solution for registering for results coming from 3rd party Activities as AndroidX's https://developer.android.com/training/basics/intents/result.

- Create `AppContextActivityResultCallback` that mirrors https://developer.android.com/reference/androidx/activity/result/ActivityResultCallback and has additional parameter `launchingActivityHasBeenKilled`
- Create `AppContextActivityResultCaller` that mirrors https://developer.android.com/reference/androidx/activity/result/ActivityResultCaller to accepts `AppContextActivityResultCallback`
- Create `AppContextActivityResultRegistry` that mirros https://developer.android.com/reference/androidx/activity/result/ActivityResultRegistry, but does not throw upon too late initialization. The original class forces registration in `onCreate` lifecycle method and when using `ReactNative` we are not able to hook into that method properly. Moreover the original class is coupled with Activity's lifecycle and we need something that would outlive Activity (and is hooked into `AppContext` lifecycle)
- Create `ActivityResultsManager` that is encapsulating all `Activity`-related-for-result-registration outside `AppContext`

bbarthec committed:
[amc][Android] Expose `registerForActivityResult` and `launchForActivityResult` via `AppContext`

bbarthec committed:
[emc][Android] Create `ReactActivityProvider`

bbarthec committed:
[emc][Android] Implement `ReactActivityProvider` in `AppContext`

bbarthec committed:
[emc][Android] Replace `ReactActivityProvider` with `CurrentActivityProvider`

bbarthec committed:
[skip ci] Update CHANGELOG

bbarthec committed:
[emc][Android] Add missing `launchForActivityResult` to the `AppContextActivityResultCaller`

bbarthec committed:
[emc][Android] Cleanup and format code

bbarthec committed:
[e-m-c] Wait with dispatching `onActivityResult` for `ReactContext` initialisation

bbarthec committed:
[e-m-c] Refactor `registerForActivityResult` API to propoerly handle Activity reacreation

- create `AppCompatActivityAware` mechanism for detecting propoerly moment when `Activity` is firstly created (similar to `onCreate` method)
- convert `registerForActivityResult` API from single-callback based to two-callback based
  - main callback is straghtforward callback that is used when the Activity outlives the process of launching other 3rd party Activity
  - secodnary (fallback) callback is stripped from RN context (no promise, etc) and only handles the scanerio when the Activity is recreated, it preserves the optional options though
- create initial API for serialization mechanism that would help preserve additional options during Activity recreation
- adapt `AppContext` to these changes
  - postpone registering modules to the very end of `AppContext` creation as the previous approach would crash now

bbarthec committed:
[emc][Android] Create `Bundleable` interface

This is an interface that would allow persisting data in between Activity destruction and recreation

bbarthec committed:
[emc][Android] Early version of `DataPersistor`

A helper class that is responsince for storing and retrieving data using `SharedPrefernces`

bbarthec committed:
[emc][Android] Adapt `Bundleable`

bbarthec committed:
[emc][Android] Cleanup `ActivityResultRegistry`, adapt `Bundleable` and Activity recreation

- cleanup, unify and comment data structures
- cleanup both code flows
  - lifecycle-based flow
  - `onActivityResult` (aka `dispatch`) flow
- adapt `DataPersistor`
- remove unused code

bbarthec committed:
[emc][Android] Remove `Bundleable` in favour of `Serializable`

bbarthec committed:
[emc][Android] Implement `DataPersistor` logic

Implement pairs of methods (setter and getter) for each supported data kind

bbarthec committed:
[emc] Improve comments

bbarthec committed:
[emc] `:expo-modules-core:spotlessApply`

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityresult/DataPersistor.kt:
```diff
@@ -0,0 +1,143 @@
+  private val accumulator = Bundle()
+
+  @Suppress("PropertyName")
+  private var _retrievedData: Bundle? = null
+  private val retrievedData: Bundle
+    get() = _retrievedData ?: retrieveData().also { _retrievedData = it }
```
```suggestion
  private val retrievedData by lazy { retrieveData() }
```
It should work in the same way ;)

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityresult/DataPersistor.kt:
```diff
@@ -0,0 +1,143 @@
+    return retrievedData.getStringArrayList(key)
+  }
+
+  fun addStringToIntMap(key: String, value: Map<String, Int>): DataPersistor {
+    accumulator.putBundle(key, bundleOf(*value.toList().toTypedArray()))
+    return this
+  }
```
```suggestion
  fun addStringToIntMap(key: String, value: Map<String, Int>): DataPersistor = apply {
    accumulator.putBundle(key, bundleOf(*value.toList().toTypedArray()))
  }
```

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityresult/DataPersistor.kt:
```diff
@@ -0,0 +1,143 @@
+    }
+  }
+
+  fun addStringToSerializableMap(key: String, value: Map<String, Serializable>): DataPersistor {
+    accumulator.putBundle(
+      key,
+      bundleOf(
+        *value
+          .toList()
+          .toTypedArray()
+      )
+    )
+    return this
+  }
```
```suggestion
  fun addStringToSerializableMap(key: String, value: Map<String, Serializable>): DataPersistor = apply {
    accumulator.putBundle(
      key,
      bundleOf(
        *value
          .toList()
          .toTypedArray()
      )
    )
  }
```

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityresult/DataPersistor.kt:
```diff
@@ -0,0 +1,143 @@
+      }
+  }
+
+  fun addBundle(key: String, value: Bundle): DataPersistor {
+    accumulator.putBundle(key, value)
+    return this
+  }
```
```suggestion
  fun addBundle(key: String, value: Bundle): DataPersistor = apply {
    accumulator.putBundle(key, value)
  }
```

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityresult/DataPersistor.kt:
```diff
@@ -0,0 +1,143 @@
+    return retrievedData.getBundle(key)
+  }
+
+  fun addSerializable(key: String, value: Serializable): DataPersistor {
+    accumulator.putSerializable(key, value)
+    return this
+  }
```
```suggestion
  fun addSerializable(key: String, value: Serializable): DataPersistor = apply {
    accumulator.putSerializable(key, value)
  }
```

lukmccall commented on packages/expo/android/src/main/java/expo/modules/ReactActivityDelegateWrapper.kt:
```diff
@@ -142,6 +144,29 @@ class ReactActivityDelegateWrapper(
   override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+    /**
+     * Workaround for a problem when results from [onActivityResult] are not properly delivered to modules.
+     * It happens when Android kills the [Activity] upon low memory scenario and recreates its later on.
```
```suggestion
     * It happens when Android kills the [Activity] upon low memory scenario and recreates it later on.
```

lukmccall reviewed (commented):
Looks very solid 🔥🔥🔥Great job 🏅

Kudo commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityaware/AppCompatActivityAwareHelper.kt:
```diff
@@ -0,0 +1,42 @@
+    val a = activity.get()
+    if (a != null) {
+      listener.onActivityAvailable(a)
+    }
```
is it intentional to emit the event in the adding method? if yes with this side effect, please add some comments.

bbarthec replied:
Yes, it is intentional. I've adapted this class from https://android.googlesource.com/platform/frameworks/support/+/HEAD/activity/activity/src/main/java/androidx/activity/contextaware/ContextAwareHelper.java and in `AndroidX` they do it that way, but yeah, I'll add some comment 😉

bbarthec replied:
I've added a comment to the `AppCompatActivityAware` interface 😉

Kudo reviewed (approved):
just leaving a nit, other than that the pr looks good to me.

bbarthec commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/AppContext.kt:
```diff
@@ -185,6 +199,11 @@ class AppContext(
 
   fun onHostResume() {
+    activityResultsManager.onHostResume(
+      requireNotNull(currentActivity) {
+        "Current Activity is not available at this moment"
+      }
```
This would actually never crash, because whenever any of lifecycle method is called (except for `OnCreate`) the current `Activity`  must be available, because it is actually propagating these events 😉 
I went with `requireNotNull` to get rid of `nullability` (I could have gone with `!!`, but we don't want that, do we? 👀).
I can go with some other message, like: `"This would never happen, but if it does, something is broken very much"`

lukmccall replied:
I think you can leave it as it is.

bbarthec committed:
[emc] Apply review suggestions

bbarthec commented on packages/expo/android/src/main/java/expo/modules/ReactActivityDelegateWrapper.kt:
```diff
@@ -142,6 +144,29 @@ class ReactActivityDelegateWrapper(
+     *
+     * TODO (@bbarthec): fix it upstream?
+     */
+    if (delegate.reactInstanceManager.currentReactContext == null) {
+      val reactContextListener = object : ReactInstanceEventListener {
+        override fun onReactContextInitialized(context: ReactContext?) {
+          delegate.reactInstanceManager.removeReactInstanceEventListener(this)
+          delegate.onActivityResult(requestCode, resultCode, data)
+        }
+      }
+      return delegate.reactInstanceManager.addReactInstanceEventListener(reactContextListener)
+    }
```
Should I add this as a CHANGELOG entry in `packages/expo`? 🤔 I'm not sure what is our approach towards `expo` package, cc @tsapeta

Kudo replied:
it's worth to add a CHANGELOG because it fixes a real bug IMO.

tsapeta replied:
I agree with Kudo 👍

bbarthec committed:
Apply review suggestions

bbarthec committed:
[emc] Fix multiple call to removed listener

It was happening because listener was added to the listeners pool adter it was called with the value. If it tires to unregister itself in its body it was happening before it was even added to the registered listeners pool.

lukmccall commented on packages/expo-modules-core/android/src/main/java/expo/modules/kotlin/activityresult/DataPersistor.kt:
```diff
@@ -0,0 +1,134 @@
+import kotlin.collections.ArrayList
+
+const val EXPIRE_KEY = "expire"
+const val EXPIRATION_TIME = 5 * 60 * 1000 // 5 min
```
```suggestion
const val EXPIRATION_TIME = 5 * DateUtils.MINUTE_IN_MILLIS
```
Up yo you ;)

bbarthec committed:
[skip ci] Apply code suggestion
